src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -407,5 +407,48 @@ ExpressionSyntax asExpression(SyntaxNode node)
                 return expr;
             }
         }
+
+        private sealed class TopLevelNullabilityRetriever : BoundTreeWalker
+        {
+            private Dictionary<SyntaxNode, TypeSymbolWithAnnotations> _map;
+            private TopLevelNullabilityRetriever(Dictionary<SyntaxNode, TypeSymbolWithAnnotations> map)
+            {
+                _map = map;
+            }
+
+            public static Dictionary<SyntaxNode, TypeSymbolWithAnnotations> BuildMap(BoundNode rewritten)
+            {
+                var map = new Dictionary<SyntaxNode, TypeSymbolWithAnnotations>();
+                new TopLevelNullabilityRetriever(map).Visit(rewritten);
+                return map;
+            }
+
+            protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpression node)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                // For the purposes of verifying types, implicit conversions should not win over lower nodes
+                // with the same syntax, as when testing we're usually testing the kind before the conversion,
+                // and we have no way from syntax of getting the correct node.
+                if (node is BoundConversion conv && conv.Syntax == conv.Operand.Syntax && conv.Operand.Kind != BoundKind.Lambda)
+                {
+                    _map[conv.Syntax] = new TypeWithState(conv.Type, conv.TopLevelNullability.FlowState.ToInternalFlowState()).ToTypeSymbolWithAnnotations();
+                    base.Visit(node);
+                }
+                else
+                {
+                    base.Visit(node);
+                    if (node is BoundExpression expr)
+                    {
+                        _map[expr.Syntax] = new TypeWithState(expr.Type, expr.TopLevelNullability.FlowState.ToInternalFlowState()).ToTypeSymbolWithAnnotations();
```

Comment:
>new TypeWithState(expr.Type, expr.TopLevelNullability.FlowState.ToInternalFlowState()).ToTypeSymbolWithAnnotations() [](start = 44, length = 116)

Consider extracting a local function.

Reply:
This visitor isn't intended to survive past the introduction of the public API. I'll leave this as is.

---
In reply to: [266133059](https://github.com/dotnet/roslyn/pull/34017#discussion_r266133059) [](ancestors = 266133059)